pkg/gcp/plugin.go

```diff
@@ -56,31 +81,77 @@ type PluginSPIImpl struct{}
 
 	sa, err := gcp.GetCredentialsConfigFromJSON([]byte(credentialsConfigJSON))
 	if err != nil {
-		return ctx, nil, fmt.Errorf("could not get service account. err: %w", err)
+		return ctx, nil, fmt.Errorf("could not get service account from %q field: %w", credentialKey, err)
+	}
+
+	if sa.Type == gcp.ServiceAccountCredentialType {
+
+		fields := map[string]string{}
+		if err := json.Unmarshal([]byte(credentialsConfigJSON), &fields); err != nil {
+			return ctx, nil, fmt.Errorf("failed to unmarshal '%q' field: %w", credentialKey, err)
+		}
+
+		for f := range fields {
+			if _, ok := serviceAccountAllowedFields[f]; !ok {
+				return ctx, nil, fmt.Errorf("forbidden fields are present. Allowed fields are %s", strings.Join(slices.Collect(maps.Keys(serviceAccountAllowedFields)), ", "))
+			}
+		}
+		jwt, err := google.JWTConfigFromJSON([]byte(credentialsConfigJSON), compute.CloudPlatformScope)
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot parse serviceAccountJSON secret value: %w", err)
+		}
+		clientOption := option.WithTokenSource(jwt.TokenSource(ctx))
+		computeService, err := compute.NewService(ctx, clientOption)
+		if err != nil {
+			return nil, nil, err
+		}
+		return ctx, computeService, nil
+
+	} else if sa.Type == gcp.ExternalAccountCredentialType {
+		err := validateExtAccountFields(sa)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid secret. Err: %w", err)
+		}
+		creds, err := google.CredentialsFromJSONWithParams(ctx, []byte(credentialsConfigJSON), google.CredentialsParams{
+			Scopes: []string{compute.CloudPlatformScope},
+		})
+		if err != nil {
+			return nil, nil, fmt.Errorf("cannot parse serviceAccountJSON secret value: %w", err)
+		}
+		clientOption := option.WithTokenSource(creds.TokenSource)
+		computeService, err := compute.NewService(ctx, clientOption)
+		if err != nil {
+			return nil, nil, err
+		}
+		return ctx, computeService, nil
+
+	} else {
+		return ctx, nil, fmt.Errorf("forbidden credential type %q used. Only %q or %q is allowed", sa.Type, gcp.ServiceAccountCredentialType, gcp.ExternalAccountCredentialType)
 	}
+}
 
-	if sa.Type != gcp.ServiceAccountCredentialType {
-		return ctx, nil, fmt.Errorf("forbidden credential type %q used. Only %q is allowed", sa.Type, gcp.ServiceAccountCredentialType)
+func validateExtAccountFields(sa *gcp.CredentialsConfig) error {
+	if strings.TrimSpace(sa.TokenURL) != allowedTokenURL {
+		return fmt.Errorf("invalid token URL found")
 	}
 
-	creds, err := google.CredentialsFromJSONWithParams(ctx, []byte(credentialsConfigJSON), google.CredentialsParams{
-		Scopes: []string{compute.CloudPlatformScope},
-	})
-	if err != nil {
-		return nil, nil, fmt.Errorf("cannot parse serviceAccountJSON secret value: %w", err)
+	if !allowedServiceAccountImpersonationURLRegExp.MatchString(sa.ServiceAccountImpersonationURL) {
+		return fmt.Errorf("invalid service_account_impersonation_url found in secret")
```

Comment:
Since `service_account_impersonation_url` is optional we need to validate it only if it is set.

```suggestion
	if len(sa.ServiceAccountImpersonationURL) > 0 && !allowedServiceAccountImpersonationURLRegExp.MatchString(sa.ServiceAccountImpersonationURL) {
		return fmt.Errorf("invalid service_account_impersonation_url: should match reg expression "+allowedServiceAccountImpersonationURLRegExp.String())
```